Source/SwiftLintFramework/Rules/Style/NoReturnVoidRule.swift

```diff
@@ -0,0 +1,258 @@
+                    }
+                }
+                func otherVoid() -> Void {}
+            }
+            """,
+            """
+            func test() {
+                guard conditionIsTrue else {
+                    sideEffects()
+                    return // comment
+                }
+                guard otherCondition else {
+                    return↓ assertionfailure("")
+                }
+                differentSideEffect()
+            }
+            """,
+            """
+            func test() {
+                guard otherCondition else {
+                    return↓ assertionfailure(""); // comment
+                }
+                differentSideEffect()
+            }
+            """,
+            """
+            func test() {
+              if x {
+                return↓ foo()
+              }
+              bar()
+            }
+            """,
+            """
+            func test() {
+              switch x {
+                case .a:
+                  return↓ foo() // return to skip baz()
+                case .b:
+                  bar()
+              }
+              baz()
+            }
+            """,
+            """
+            func test() {
+              if check {
+                if otherCheck {
+                  return↓ foo()
+                }
+              }
+              bar()
+            }
+            """,
+            """
+            func test() {
+                return↓ foo()
+            }
+            """,
+            """
+            func test() {
+              return foo({
+                return bar()
+              })
+              return↓ foo()
+            }
+            """,
+            """
+            func test() {
+              guard x else {
+                return↓ foo()
+              }
+              bar()
+            }
+            """,
+            """
+            func test() {
+              let closure: () -> () = {
+                return assert()
+              }
+              if check {
+                if otherCheck {
+                  return // comments are fine
+                }
+              }
+              return↓ foo()
+            }
+            """
+        ]
+    )
+
+    public func validate(file: File, kind: SwiftDeclarationKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        guard SwiftDeclarationKind.functionKinds.contains(kind),
+            dictionary.isVoid(),
+            !dictionary.isInit(),
+            let bodyRange = dictionary.byteRange()
+        else {
+            return []
+        }
+
+        let statementRanges = dictionary.statementRanges()
+        var bodyRangeAfterLastStatement = bodyRange
+
+        if let lastStatement = statementRanges?.last {
+            bodyRangeAfterLastStatement.location = lastStatement.upperBound
+        }
+
+        let ranges: [NSRange] = [bodyRangeAfterLastStatement] + (statementRanges ?? [])
```

Comment:
Suggested changes:

```swift
let ranges = [bodyRangeAfterLastStatement] + (statementRanges ?? [])
return ranges.compactMap { range in
    let nonCommentokens = file.syntaxMap.tokens.lazy
        .filter { range.contains($0.offset) && !$0.isComment }
    guard let last = nonCommentokens.last,
        let lastReturn = nonCommentokens.last(where: { $0.isReturnKeyword(in: file) }),
        last != lastReturn
    else {
        return nil
    }

    let location = Location(file: file, byteOffset: lastReturn.offset)
    return StyleViolation(ruleDescription: type(of: self).description,
                          severity: configuration.severity,
                          location: location)
}
```

This should be more efficient by virtue of not mapping then compact mapping, using `lazy` to avoid computing all filtered tokens that aren't comments since we just care about the last one(s).

It also changes the violation location to be at the beginning of the `return` keyword, which looks a bit nicer IMHO.

Reply:
Love it! Yeah looks like the rule was getting a little slow. 👍